compiler/rustc_session/src/options.rs

```diff
@@ -132,7 +132,7 @@ top_level_options!(
         lint_cap: Option<lint::Level> [TRACKED],
         force_warns: Vec<String> [TRACKED],
         describe_lints: bool [UNTRACKED],
-        output_types: OutputTypes [TRACKED],
+        output_types: OutputTypes [TRACKED_NO_CRATE_HASH],
```

Comment:
I am concerned about the fact that changing which output types are requested doesn't change the crate hash due to `if output_types.should_codegen()` occurring in several locations that can affect the crate metadata like the list of reachable non-generics, the list of exported symbols and `dependency_format::calculate_type`. In principle these `should_codegen()` fast-paths could be removed, but it will likely regress `cargo check` performance.

Reply:
In https://internals.rust-lang.org/t/rfc-emit-check/14711/5?u=jsgf I measured it as ~8% difference on `cargo check` for Cargo itself. I don't know how that relates to anything else.

I experimented with making a distinction between `--emit metadata` and `--emit check` and corresponding `.rmeta` and `.rcheck` files, but the addition of a new file type in the mix made things complex in location resolution in rustc and in Cargo itself. It would be much easier to just make `.rmeta` invariant.

One advantage, in principle, is that if `cargo check`'s emitted metadata is "full fat", then it can be used immediately from cache in `cargo build` which would effectively make the dependency graph flat and allow the rlibs to be generated completely in parallel - though I expect that would only really be interesting for builds which are closer to build from scratch vs incremental rebuild.

Reply:
I also wonder what the practical effect of this will be. I presume the risk is that some build system - like `cargo` will intermix invocations of `--emit metadata` (cargo check) and `--emit link,metadata` (cargo build) and get failures if check metadata gets used for builds.

But what's the actual outcome? In the worst case, it could appear to succeed but generate subtly broken output. But I don't think that's the case here, is it? It will fail to find " list of reachable non-generics, the list of exported symbols and dependency_format::calculate_type", and fail as a result. Vs if the hashes are different, it will fail for that reason ("found possibly newer version of crate ...").

(If need be, could we extend the .rmeta format to include a type to explicitly flag "for check use only"?)

I haven't looked into how `cargo` specifically managed check vs build .rmetas, but I think it avoids the problem by always regenerating .rmetas as part of a build. In my own experiments with Buck, I make sure to put them in separate dirs (which is the motivation for this PR, since currently doing that changes the crate hash independent of anything else).

Either way, any failure is a bug in the build system, not something that an end-user would expect to see with a correctly implemented build system.

Reply:
> I also wonder what the practical effect of this will be. I presume the risk is that some build system - like cargo will intermix invocations of --emit metadata (cargo check) and --emit link,metadata (cargo build) and get failures if check metadata gets used for builds.

That is exactly what you want to do, right? Using the `--emit metadata` output of dependencies to compile `--emit link` crates.

> But what's the actual outcome? In the worst case, it could appear to succeed but generate subtly broken output. But I don't think that's the case here, is it? It will fail to find " list of reachable non-generics, the list of exported symbols and dependency_format::calculate_type", and fail as a result. Vs if the hashes are different, it will fail for that reason ("found possibly newer version of crate ...").

A wrong result of dependency_format::calculate_type will cause linking to silently fail due to mixing up static and dynamic linking of crates. For example trying to statically link a crate that is already linked into a dynamic library dependency.

A wrong result for the list of reachable non-generics and exported symbols would likely result in the mono item collector creating mono items that are already included in dependencies (or maybe omitting necessary ones), causing duplicate linker errors.

> I haven't looked into how cargo specifically managed check vs build .rmetas, but I think it avoids the problem by always regenerating .rmetas as part of a build.

It currently handles check and regular builds independently, but we want `cargo check` to reuse `cargo build` rmeta files in the future.

Reply:
Yes, just the types being hashes is fine. `-Zno-codegen` doesn't affect the metadata in anyway AFAIK, so not hashing it for the crate hash is fine.

Reply:
> That is exactly what you want to do, right? Using the --emit metadata output of dependencies to compile --emit link crates.

If `--emit link` and `--emit metadata` inherently generate different crate hashes, then that's useless to me. I currently have to use `--emit link -Zno-codegen` to generate complete hash-compatible metadata. That's fine for experiments, but I'd like to get to something that can be stabilized, and ideally directly expresses the intent.

My current opinion is that using `--emit metadata` which always generates the same metadata regardless of other emits which is crate-hash compatible with `--emit link` would be ideal, but I'm open to any suggestion that meets the goals of:
- separate rustc invocations to generate compatible metadata and codegen
- on a stabliization track

Reply:
> My current opinion is that using --emit metadata which always generates the same metadata regardless of other emits which is crate-hash compatible with --emit link would be ideal

I agree, but probably not everyone may agree with the associated perf loss unfortunately.

Reply:
> the associated perf loss

Just a question, is it possible to reuse these fat rmeta generated in checking (seems it contains MIR and type information) when building binaries? If so, we can save 100% of check time (the time populating rmeta) during the build.
But sadly we currently just re-work on analysis and re-generate rmeta again when `--emit link`.

Reply:
Yes, build -> check reusing is already possible, but simply not implemented in cargo. There should be an issue open for that. check -> build reusing won't work even with this PR.

Reply:
Is check -> build reusing technically possible? I'm not sure if everything needed for codegen is already available in rmeta.

Reply:
In theory it would be possible to never exclude anything from the rmeta even when no codegen is required. This would allow using the check rmeta for builds, but still require all crates to be compiled with codegen. The benefit of this would be more parallelization as codegen for all crates can immediately start without having to wait for the compilation of dependencies to get metadata. The downside is that the check metadata becomes larger which slows down check runs.